File: insertdeletegetrandom01.js
//Objective is to design a data structure that supports inserting, deleting, and 
//getting a random value all in O(1) time.


//As stated in the question, everything here is O(1), so we use a 
//hashmap and an array to keep track of our values.
class Set {
    constructor() {
        this.map = {}
        this.arr = []
    }

    //Here, we update the map first since the index of the value on the map
    //should always be length - 1
    insert(val) {
        if (val in this.map) {
            return false 
        }

        this.map[val] = this.arr.length 
        this.arr.push(val)

        return true
    }

    //Here, we swap the values in the array and map for the value we want 
    //to delete. Then, we pop it out of the array and remove it from the map,
    //both of which are O(1)
    delete(val) {
        if (!(val in this.map)) {
            return false
        }

        let end = this.map[val]
        let original = this.arr[this.arr.length - 1]
        this.map[original] = end 
        this.arr[end] = original 
        delete this.map[val]
        this.arr.pop()

        return true
    }

    //Here, we get a random value in the array using Math.random()
    getRandom() {
        return this.arr[Math.floor(Math.random() * this.arr.length)]
    }
}

let set = new Set()
set.insert(5)
set.insert(10)
set.delete(5)

File: insertdeletegetrandom01.java
//Java Solution

class RandomizedSet {
    
    Map<Integer, Integer> map;
    List<Integer> arr;
    Random rand = new Random();
    /** Initialize your data structure here. */
    public RandomizedSet() {
        map = new HashMap();
        arr = new ArrayList();
    }
    
    /** Inserts a value to the set. Returns true if the set did not already contain the specified element. */
    public boolean insert(int val) {
        if (map.containsKey(val)) {
            return false;
        }
        
        map.put(val, arr.size());
        arr.add(arr.size(), val);
        return true;
    }
    
    /** Removes a value from the set. Returns true if the set contained the specified element. */
    public boolean remove(int val) {
        if (!map.containsKey(val)) {
            return false;
        }
        
        int original = map.get(val);
        int end = arr.get(arr.size() - 1);
        map.put(end, original);
        arr.set(original, end);
        map.remove(val);
        arr.remove(arr.size() - 1);
        
        return true;
    }
    
    /** Get a random element from the set. */
    public int getRandom() {
        return arr.get(rand.nextInt(arr.size()));
    }
}
